File: app/src/main/java/com/keylesspalace/tusky/PreferencesActivity.java
/* Copyright 2017 Andrew Dawson
 *
 * This file is a part of Tusky.
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the
 * GNU General Public License as published by the Free Software Foundation; either version 3 of the
 * License, or (at your option) any later version.
 *
 * Tusky is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
 * Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with Tusky; if not,
 * see <http://www.gnu.org/licenses>. */

package com.keylesspalace.tusky;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.annotation.XmlRes;
import android.support.v7.app.ActionBar;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;

import com.keylesspalace.tusky.fragment.PreferencesFragment;
import com.keylesspalace.tusky.util.ThemeUtils;

public class PreferencesActivity extends BaseActivity
        implements SharedPreferences.OnSharedPreferenceChangeListener {

    private boolean restartActivitiesOnExit;
    private @XmlRes int currentPreferences;
    private @StringRes int currentTitle;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (savedInstanceState != null) {
            restartActivitiesOnExit = savedInstanceState.getBoolean("restart");
        } else {
            Bundle extras = getIntent().getExtras();
            restartActivitiesOnExit = extras != null && extras.getBoolean("restart");
        }

        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(this);

        setContentView(R.layout.activity_preferences);

        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setDisplayShowHomeEnabled(true);
        }

        preferences.registerOnSharedPreferenceChangeListener(this);

        if(savedInstanceState == null) {
            currentPreferences = R.xml.preferences;
            currentTitle = R.string.action_view_preferences;
        } else {
            currentPreferences = savedInstanceState.getInt("preferences");
            currentTitle = savedInstanceState.getInt("title");
        }
        showFragment(currentPreferences, currentTitle);

    }

    public void showFragment(@XmlRes int preferenceId, @StringRes int title) {

        //TODO: cache the Fragments so they can be reused
        getFragmentManager().beginTransaction()
                .replace(R.id.fragment_container, PreferencesFragment.newInstance(preferenceId))
                .commit();

        getFragmentManager().executePendingTransactions();

        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setTitle(title);
        }

        currentPreferences = preferenceId;
        currentTitle = title;
    }

    private void saveInstanceState(Bundle outState) {
        outState.putBoolean("restart", restartActivitiesOnExit);
        outState.putInt("preferences", currentPreferences);
        outState.putInt("title", currentTitle);
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        saveInstanceState(outState);
        super.onSaveInstanceState(outState);
    }

    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        switch (key) {
            case "appTheme": {
                String theme = sharedPreferences.getString("appTheme", TuskyApplication.APP_THEME_DEFAULT);
                ThemeUtils.setAppNightMode(theme);
                restartActivitiesOnExit = true;

                // recreate() could be used instead, but it doesn't have an animation B).
                Intent intent = getIntent();
                intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_NEW_TASK);
                Bundle savedInstanceState = new Bundle();
                saveInstanceState(savedInstanceState);
                intent.putExtras(savedInstanceState);
                startActivity(intent);
                finish();
                overridePendingTransition(R.anim.fade_in, R.anim.fade_out);
            }
            case "statusTextSize": {
                restartActivitiesOnExit = true;
                break;
            }
            case "notificationsEnabled": {
                boolean enabled = sharedPreferences.getBoolean("notificationsEnabled", true);
                if (enabled) {
                    enablePushNotifications();
                } else {
                    disablePushNotifications();
                }
                break;
            }
            case "pullNotificationCheckInterval": {
                String s = sharedPreferences.getString("pullNotificationCheckInterval", "15");
                long minutes = Long.valueOf(s);
                setPullNotificationCheckInterval(minutes);
                break;
            }
        }
    }

    @Override
    public void onBackPressed() {
        //if we are not on the top level, show the top level. Else exit the activity
        if(currentPreferences != R.xml.preferences) {
            showFragment(R.xml.preferences, R.string.action_view_preferences);

        } else {
        /* Switching themes won't actually change the theme of activities on the back stack.
         * Either the back stack activities need to all be recreated, or do the easier thing, which
         * is hijack the back button press and use it to launch a new MainActivity and clear the
         * back stack. */
            if (restartActivitiesOnExit) {
                Intent intent = new Intent(this, MainActivity.class);
                intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
                startActivity(intent);
            } else {
                super.onBackPressed();
            }
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home: {
                onBackPressed();
                return true;
            }
        }
        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/com/keylesspalace/tusky/PreferencesActivity.kt
/* Copyright 2018 Conny Duck
 *
 * This file is a part of Tusky.
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the
 * GNU General Public License as published by the Free Software Foundation; either version 3 of the
 * License, or (at your option) any later version.
 *
 * Tusky is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
 * Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with Tusky; if not,
 * see <http://www.gnu.org/licenses>. */

package com.keylesspalace.tusky

import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.preference.PreferenceManager
import androidx.fragment.app.Fragment
import android.util.Log
import android.view.MenuItem
import com.keylesspalace.tusky.appstore.EventHub
import com.keylesspalace.tusky.appstore.PreferenceChangedEvent
import com.keylesspalace.tusky.fragment.preference.*
import com.keylesspalace.tusky.util.ThemeUtils
import com.keylesspalace.tusky.util.getNonNullString
import dagger.android.AndroidInjector
import dagger.android.DispatchingAndroidInjector
import dagger.android.support.HasSupportFragmentInjector
import kotlinx.android.synthetic.main.toolbar_basic.*
import java.lang.IllegalArgumentException
import javax.inject.Inject
import androidx.appcompat.app.AppCompatDelegate

class PreferencesActivity : BaseActivity(), SharedPreferences.OnSharedPreferenceChangeListener, HasSupportFragmentInjector {

    @Inject
    lateinit var eventHub: EventHub

    @Inject
    lateinit var fragmentInjector: DispatchingAndroidInjector<Fragment>

    private var restartActivitiesOnExit: Boolean = false

    override fun onCreate(savedInstanceState: Bundle?) {

        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_preferences)

        setSupportActionBar(toolbar)
        supportActionBar?.run {
            setDisplayHomeAsUpEnabled(true)
            setDisplayShowHomeEnabled(true)
        }

        val fragment: Fragment = when(intent.getIntExtra(EXTRA_PREFERENCE_TYPE, 0)) {
            GENERAL_PREFERENCES -> {
                setTitle(R.string.action_view_preferences)
                PreferencesFragment.newInstance()
            }
            ACCOUNT_PREFERENCES -> {
                setTitle(R.string.action_view_account_preferences)
                AccountPreferencesFragment.newInstance()
            }
            NOTIFICATION_PREFERENCES -> {
                setTitle(R.string.pref_title_edit_notification_settings)
                NotificationPreferencesFragment.newInstance()
            }
            TAB_FILTER_PREFERENCES -> {
                setTitle(R.string.pref_title_status_tabs)
                TabFilterPreferencesFragment.newInstance()
            }
            PROXY_PREFERENCES -> {
                setTitle(R.string.pref_title_http_proxy_settings)
                ProxyPreferencesFragment.newInstance()
            }
            else -> throw IllegalArgumentException("preferenceType not known")
        }

        supportFragmentManager.beginTransaction()
                .replace(R.id.fragment_container, fragment)
                .commit()

        restartActivitiesOnExit = intent.getBooleanExtra("restart", false)

    }

    override fun onResume() {
        super.onResume()
        PreferenceManager.getDefaultSharedPreferences(this).registerOnSharedPreferenceChangeListener(this)
    }

    override fun onPause() {
        super.onPause()
        PreferenceManager.getDefaultSharedPreferences(this).unregisterOnSharedPreferenceChangeListener(this)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                onBackPressed()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private fun saveInstanceState(outState: Bundle) {
        outState.putBoolean("restart", restartActivitiesOnExit)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putBoolean("restart", restartActivitiesOnExit)
        super.onSaveInstanceState(outState)
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
        when (key) {
            "appTheme" -> {
                val theme = sharedPreferences.getNonNullString("appTheme", ThemeUtils.APP_THEME_DEFAULT)
                Log.d("activeTheme", theme)
                ThemeUtils().setAppNightMode(theme, this)
                restartActivitiesOnExit = true

                // recreate() could be used instead, but it doesn't have an animation B).
                intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_NEW_TASK
                val savedInstanceState = Bundle()
                saveInstanceState(savedInstanceState)
                intent.putExtras(savedInstanceState)
                startActivityWithSlideInAnimation(intent)
                finish()
                overridePendingTransition(R.anim.fade_in, R.anim.fade_out)

                // MODE_NIGHT_FOLLOW_SYSTEM workaround part 2 :/
                when(theme){
                    ThemeUtils.THEME_SYSTEM -> {
                        AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM)
                    }
                }
                //workaround end

            }
            "statusTextSize" -> {
                restartActivitiesOnExit = true
            }
            "absoluteTimeView" -> {
                restartActivitiesOnExit = true
            }
        }

        eventHub.dispatch(PreferenceChangedEvent(key))
    }


    override fun onBackPressed() {
        /* Switching themes won't actually change the theme of activities on the back stack.
         * Either the back stack activities need to all be recreated, or do the easier thing, which
         * is hijack the back button press and use it to launch a new MainActivity and clear the
         * back stack. */
        if (restartActivitiesOnExit) {
            val intent = Intent(this, MainActivity::class.java)
            intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
            startActivityWithSlideInAnimation(intent)
        } else {
            super.onBackPressed()
        }
    }

    override fun supportFragmentInjector(): AndroidInjector<Fragment> {
        return fragmentInjector
    }

    companion object {

        const val GENERAL_PREFERENCES = 0
        const val ACCOUNT_PREFERENCES = 1
        const val NOTIFICATION_PREFERENCES = 2
        const val TAB_FILTER_PREFERENCES = 3
        const val PROXY_PREFERENCES = 4
        private const val EXTRA_PREFERENCE_TYPE = "EXTRA_PREFERENCE_TYPE"

        @JvmStatic
        fun newIntent(context: Context, preferenceType: Int): Intent {
            val intent = Intent(context, PreferencesActivity::class.java)
            intent.putExtra(EXTRA_PREFERENCE_TYPE, preferenceType)
            return intent
        }
    }

}
